merkle_tree/src/main.rs

```diff
@@ -0,0 +1,432 @@
+use sha256::digest;
+
+struct Merkle_Tree {
+    tree: Vec<String>,
+    leaves: usize,
+    og_leaves: Vec<String>,
+}
+/* */
+impl Merkle_Tree {
+    /// Builds the tree concatenating the diferent levels into one array.
+    /// The final array tree goes from root to leaves -> [root....branch_n, branch_n+1....leaf].
+    fn new(data_vector: Vec<&str>) -> Self {
+        let string_data_vec: Vec<String> = data_vector.iter().map(|s| s.to_string()).collect();
+        let pow2_data_vector = extend_to_power2_size(&string_data_vec);
+        let mut level_n = vec![];
+        for data in &pow2_data_vector {
+            level_n.push(digest(data.clone()));
+        }
+        let mut current_tree = vec![];
+        for level in 0..(pow2_data_vector.len() as f64).sqrt().ceil() as usize {
+            //Amount of levels
+            let mut level_n_minus1 = vec![];
+            for i in (0..level_n.len()).step_by(2) {
+                //Build new level from the previous one
+                level_n_minus1.push(digest(level_n[i].clone() + &level_n[i + 1]));
+            }
+            let next_current_level = level_n_minus1.clone();
+            if level == 0 {
+                level_n_minus1.append(&mut level_n); //Concatenating the leaves.
+            } else {
+                level_n_minus1.append(&mut current_tree); //Concatenating the previous levels.
+            }
+            current_tree = level_n_minus1.clone();
+            level_n = next_current_level;
+        }
+
+        Self {
+            tree: current_tree.clone(),
+            leaves: data_vector.len(),
+            og_leaves: string_data_vec.clone(),
+        }
+    }
+
+    /// Given a proof and an index of a data hash the function must determine wether the proof is correct.
+    /// For understanding this function we can consider the following tree:
+    ///               root
+    ///          h12        h34
+    ///        h1   h2    h3   h4
+    ///        |    |     |     |
+    ///     data1 data2 data3 data4
+    /// The array representation would be: [root,h12,h34,h1,h2,h3,h4]
+    /// With the indexes                     0   1   2   3  4  5   6
+    fn verify(&self, proofs: Vec<String>, mut leaf_index: usize) -> bool {
+        let mut hash_for_verification = self.tree[leaf_index].clone();
```

Comment:
Can we avoid this clone?

Reply:
Also, keep in mind that accessing with brackets can panic, it's better to use `.get(leaf_index)`